enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,213 @@
+optional and pluggable.
+
+In the enhancement proposal we will describe an architecture which would allow
+customers to install individual compomnents as rpm packages.
+
+
+## Motivation
+
+The principal factor that gave birth to the idea of MicroShift was the need to bring
+the OpenShift deployment model into the Edge where the power, memory and CPU budgets
+are reduced.
+
+While the default configuration makes sense to most of our customers in the Device
+Edge area, still our current CSI, CNI, and router components use 35% of the memory
+footprint and a significative CPU and power footprint. Making those components
+pluggable would allow customers who don't require advanced network features,
+persistent volumes, or ingress/routes to disable or switch those technologies
+and reduce the memory/cpu and disk/over-the-wire footprint.
+
+
+### User Stories
+
+An application developer needs the whole stack of components, creates an image
+that contains:
+1. microshift
+
+Nothing would change in this case, installing microshift, which becomnes a virtual
+package, installs the necessary dependencies:
+* microshift-core
+* microshift-networking-ovn
+* microshift-storage-topolvm
+* microshift-routes
+
+as well as any other dependencies.
+
+An application developer only needs advanced networking, but no persistent volumes,
+and no routes/ingress.  Creates an image that contains:
+1. microshift-core
+2. microshift-networking-ovn
+
+
+An application developer wants the bare minimum set of components creates an image
+with:
+1. microshift-core
+
+An application developer wanting bare minimum networking but persistent volumes
+creates an image with:
+1. microshift-core
+2. microshift-storage-topolvm
+
+### Goals
+* Allow networking, storage and router core components to be optionally
+  installed.
+
+* As we refactor and re-use the kustomize package used for loading user workloads,
+  add support for loading yamls from multiple directories, i.e.:
+  * `/usr/lib/microshift/manifests.d/user-app-1/`
+  * `/usr/lib/microshift/manifests.d/user-app-2/`
+  * `/etc/microshift/manifests.d/user-app-3/`
+  * `...`
+
+* Split the RPM packages to resemble those optional components.
+
+* Preserve previous behavior in terms of RPM package installation.
+
+### Non-Goals
+
+* Making DNS component optional.
+* Making the certificate management component optional.
+
+## Proposal
+
+To allow modularity of the described core components, the following changes
+would be introduced:
+
+Update the service manager to allow core components as kustomization yamls,
+in the same way we already support for user application manifests. i.e.:
+
+`/usr/lib/microshift/components.d/networking-ovn/kustomize.yaml`
+`/usr/lib/microshift/components.d/storage-topolvm/kustomize.yaml`
+`/usr/lib/microshift/components.d/openshift-router/kustomize.yaml`
+
+  MicroShift component manager would scan the `/usr/lib/microshift/components.d/`
```

Comment:
Yes, my idea was to include as metadata (a yaml file in the directory) the components that a given component depends on, and which components depend on this one (to make extensibility easier)

Reply:
Is that necessary? I mean, MicroShift code would load these kustomize.yaml files sequentially, right? 
If so, it's only needed to decide about the order of loading, which can be done in .d style by using numbering in the directory names.

Reply:
@ggiguash I can imagine this working when we're in total control of "addons" and there's a limited number of them. If the community unleashes creativity it might be hard to keep the numbers right.

If multus wasn't rewritten and I have my facts straight, then it must start after primary cni **starts** (not is deployed), because multus just has a one-time script on daemonset start and then sleeps (rest is handled by binary on disk) - when starting it'll scan the cni dir for other CNI to use it as a delegate and it sets itself as primary (because it'll rename primary's config and create its own as `00-multus.conf`).
So multus deployment directly depends on other CNI already running, but other components might be non-blocking as in "deploy and go on", so we have already some kind of need for dependency ordering

And this might not be relevant for 4.13 since we want to keep it minimal, but about the future releases? What about upstream possibilities?

Reply:
So, yeah, we could rely on filename ordering (the classical `99_run-me-last`), but that'd need _some_ form of convention then to avoid the problem Patryk mentioned.
I'm wondering, though, in how far we actually need ordering for hosted components. For example, while we're deploying them sequentially today, services start up correctly even though OVNk takes a long time to download images and eventually start. The rest just gets blocked until then.
The multus case you're mentioned @pmtk needs some investigation, but AFAIK multus doesn't deploy manifests, so this problem would need solution on the host-level anyway, or?

Reply:
Microshift doesn't need to know the order of each plugin, but it shall have awareness of which plugin category needs be deployed first or last. Multus is implementation details which is not exposed to microshfit, it belongs to the metaplugin category. User shall be able to request a metaplugin type while starting microshift, the microshift then call the metaplugin installation at whatever step it wants, as long as it makes sure the metaplugin installation is called after the cni network plugin. The scan/overwrite of cni config file by multus shall be included in multus specific installation binary or package.